cmds-42/usr.bin/rcs/man/rcsdiff.1
.TH RCSDIFF L "October 30, 1989" "Purdue University"
.SH NAME
rcsdiff \- compare RCS revisions
.SH SYNOPSIS
\fBrcsdiff\fR [ \fB\-biwt\fR ] [ \fB\-cefhn\fR ] [ \fB\-q\fR ] [ \fB\-r\fIrev1\fR ] [ \fB\-r\fIrev2\fR ] file ...
.SH DESCRIPTION
.I Rcsdiff
runs
.IR diff (1)
to compare two revisions of each RCS file given.
A file name ending in ',v' is an RCS file name, otherwise a
working file name. \fIRcsdiff\fR derives the working file name from the RCS
file name and vice versa, as explained in
.IR co (L).
Pairs consisting
of both an RCS and a working file name may also be specified.
.PP
The options \fB\-b\fR, \fB\-i\fR, \fB\-w\fR, \fB\-t\fR,
\fB\-c\fR, \fB\-e\fR, \fB\-f\fR, and \fB\-h\fR,
have the same effect as described in
.IR diff (1);
option
\fB\-n\fR generates an edit script of the format used by RCS.
The option \fB\-q\fR suppresses diagnostic output.
.PP
If both \fIrev1\fR and \fIrev2\fR
are omitted, \fIrcsdiff\fR compares the latest revision on the
default branch (normally the highest branch on the trunk)
with the contents of the corresponding working file. This is useful
for determining what you changed since the last checkin.
.PP
If \fIrev1\fR is given, but \fIrev2\fR is omitted,
\fIrcsdiff\fR compares revision \fIrev1\fR of the RCS file with
the contents of the corresponding working file.
.PP
If both \fIrev1\fR and \fIrev2\fR are given,
\fIrcsdiff\fR compares revisions \fIrev1\fR and \fIrev2\fR of the RCS file.
.PP
Both \fIrev1\fR and \fIrev2\fR may be given numerically or symbolically,
and may actually be attached to any of the options.
.SH EXAMPLES
.nf
The command

.B "        rcsdiff  f.c

.fi
runs \fIdiff\fR on the latest revision on the default branch of RCS file f.c,v
and the contents of working file f.c.
.SH DIAGNOSTICS
The exit status is 0 if there were no differences during the last comparison,
1 if there were differences, and 2 if there were errors.
.SH IDENTIFICATION
.de VL
\\$2
..
Author: Walter F. Tichy,
Purdue University, West Lafayette, IN, 47907.
.sp 0
Revision Number:
.VL $Revision: 1.4 $
; Release Date:
.VL $Date: 89/10/30 17:36:42 $
\&.
.sp 0
Copyright \(co 1982, 1988, 1989 by Walter F. Tichy.
.SH SEE ALSO
ci(L), co(L), diff(1), ident(L), rcs(L), rcsintro(L), rcsmerge(L), rlog(L)
.br
Walter F. Tichy, "Design, Implementation, and Evaluation of a Revision Control
System," in \fIProceedings of the 6th International Conference on Software
Engineering\fR, IEEE, Tokyo, Sept. 1982.


cmds-42/ChangeLog
usr.ucb/from/from.c				mmeyer 8/22/89

Change from to do the from on geteuid(), instead of getlogin() which
looks at who is in /etc/utmp.  

usr.ucb/telnet/telnet.c				mmeyer 8/22/89

Fix for bug #2889.

bin/sed/sed1.c					mmeyer 8/21/89

Fix for bug #2831.

usr.bin/calendar/calendar.{c,sh}		mmeyer 8/21/89

Fix for bug #2794

usr.ucb/Mail					mmeyer 6/27/89

Upgraded to Tahoe sources and fixed a few bugs.

etc/mount/mount.c				mmeyer 6/26/89

Change reversal of mounting of nfs partitions.

etc/init/init.c					mmeyer 6/23/89

Changes for onerror/onoption.

etc/rlogind/rlogind.c				mmeyer 6/14/89

Took out malloc from signal handler.

etc/nu/*					mmeyer 6/13/89

Lee's new nu.

etc/ifconfig/ifconfig.c				mmeyer 6/13/89
etc/bootpd/*

Changes for sexy netinit

bin/awk/awk.g.y					mmeyer 6/1/89

Fixed core dump for Trey.  

etc/quot/quot.c					mmeyer 5/31/89

Changed acct() to print out the correct number of 1024 byte blocks.

bin/sh/homedir.c				mmeyer 5/26/89

Changed sh to use getpwnam() instead of searching through /etc/passwd

usr.ucb/indent/lexi.c				wrp    5/22/89
usr.ucb/indent/indent.c		
usr.ucb/indent/indent_globs.h

Bill Parkhurst's changes for Objective-C 4.0 support, along with
a bugfix that Brian and Morris made.


bin/diff/diffh_vers.c				mike   5/22/89
usr.bin/diff3/Makefile				

Bugfix from Mike DeMoney.  Fixes problem that occurs during integration.

usr.bin/tip/acu.c				mmeyer 5/20/89
usr.bin/tip/cu.c
usr.bin/tip/tip.c
usr.bin/tip/cmds.c

Changed references to connect(), abort() and send() to compile
with shlibs.

bin/csh/sh.c					mmeyer 5/19/89

Integrated the atexit() mod.

usr.ucb/rlogin/rlogin.c				mmeyer 5/19/89

Verified that Mike DeMoney's change to make chars with 0x7f was in.

bin/make/rules.c				mmeyer 5/19/89
bin/make/doname.c	

Integrated Mike's fixes.

usr.etc/mach_swapon/swaptab			mmeyer 5/19/89

Changed to /private/vm/swaptab instead of /etc/vm/swaptab

usr.bin/calendar/Makefile			mmeyer 5/18/89

Changed Makefile so /usr/lib/calendar would install correctly.

etc/route/route.c				mmeyer 5/16/89
etc/rwhod/rwhod.c
usr.etc/nfsstat/nfsstat.c
usr.etc/rpc.rstatd/rpc.rstatd.c
usr.ucb/sysline/sysline.c (not that this will ever run on our system)
usr.ucb/sendmail/src/conf.c
usr.bin/uucp/acucntrl.c

Changed all to nlist "/mach" instead of "/vmunix"

etc/syslog/syslog.c				mmeyer 5/15/89

Changed syslog to print out "mach:" instead of "vmunix:"

bin/date/date.c					mmeyer 5/14/89

Initialized tz.{tz_minuteswest,tz_dsttime} to zero.  This bug
cropped up when linking with shared libraries.  Kevin Enderby
has a bug report.

mk/cmds/usr/ucb/w/Makefile			mmeyer 5/14/89

Linked uptime to w in the INSTALL phase

usr.bin/uucp/fio.c				mmeyer 5/9/89
usr.bin/uucp/gio.c
usr.bin/uucp/logent.c
usr.bin/uucp/tio.c

Ifdef'd syslog() so uucp would link with the libsys_s shlib.
	

usr.bin/uucp/Makefile

Took out duplicate index.c from uucp.

usr.bin/troff/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  troff had its own 
version of calloc,realloc,malloc and free.

usr.bin/nroff/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  nroff had its own 
version of calloc,realloc,malloc and free.

usr.ucb/rusers/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  rusers had its own 
version of calloc,realloc,malloc and free.

usr.ucb/msgs/msgs.c	  			mmeyer 5/9/89

Ifdef'd send() so /usr/ucb/msgs would link with the libsys_s shlib.

usr.ucb/more/more.c	  			mmeyer 5/9/89

Ifdef'd strlen() so /usr/ucb/more would link with the libsys_s shlib.

usr.ucb/ex/exrecover.c  			mmeyer 5/9/89

Ifdef'd fprintf() so /usr/ucb/exrecover would link with the libsys_s shlib.

usr.ucb/Mail/aux.c  				mmeyer 5/9/89

Ifdef'd index() so /usr/ucb/Mail would link with the libsys_s shlib.

usr.ucb/Mail/send.c				mmeyer 5/9/89
usr.ucb/Mail/cmd1.c
usr.ucb/Mail/cmd2.c
usr.ucb/Mail/collect.c
usr.ucb/Mail/edit.c
usr.ucb/Mail/fio.c
usr.ucb/Mail/quit.c

Ifdef'd send() so /usr/ucb/Mail would link with the libsys_s shlib.

usr.ucb/Mail/head.c
usr.ucb/Mail/list.c
usr.ucb/Mail/aux.c

Ifdef'd raise() so /usr/ucb/Mail would link with the libsys_s shlib.

etc/sa/sa.c					mmeyer 5/9/89

Ifdef'd bcmp() so /etc/sa would link with the libsys_s shlib.

etc/routed/defs.h				mmeyer 5/9/89
etc/routed/startup.c
etc/routed/inet.c 

Ifdef'd inet_makeaddr() so /etc/routed would link with the libsys_s shlib.

bin/su/su.c					mmeyer 5/9/89

Unifdef'd getenv() so su would link with the libsys_s shlib.

bin/awk/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  awk had its own 
version of calloc,realloc,malloc and free.

All CMDS Makefiles				mmeyer 5/8/89

Changed all Makefiles to have -O -g for CFLAGS and -lsys_s for
LIBS.  Changes above are for commands that didn't link.

cmds/bin/cp/cp.c				mmeyer 5/8/89
cmds/bin/cat/cat.c

Integrated Avie's mapped file changes.  He claims a 10% performace
win with mapped files.

cmds/bin/make/main.c				mmeyer 5/1/89

Changed main() to have make to a setrlimit(RLIMIT_STACK) to trap
infinitely recursive macros instead of core dumping.

cmds/bin/sh/homedir.c				mmeyer  4/30/89

Changed sh to use getpwent() rather than grep'ing through /etc/passwd.

cmds/bin/tar/tar.c				mmeyer  4/30/89

Changed the "v" flag printf()'s to fprintf(stderr)'s

cmds/etc/icheck/icheck.c			mmeyer  4/30/89

Modified icheck to know about fast symlinks.

cmds/usr.ucb/rlogin/rlogin.c			mmeyer

Added the oob() signal handler for the SIGURG signal.  Data
was getting lost between the socket and the rlogin program.

cmds/bin/csh/sh.c				mmeyer

Added lines 1022 to 1029 in dosource() which check for no operand 
for source flag.

Changed lines 1036 to 1042 in dosource() to just print out an error
using perror() when an open of a source'd file fails instead of not
letting the login session continue.

cmds/bin/passwd/passwd.c			mmeyer

Took out a check against /bin/sh so a user can switch back
and forth from csh to sh and back again.


cmds-42/usr.bin/rcs/src/rcsdiff.c
/*
 *                     RCS rcsdiff operation
 */
#ifndef lint
static char rcsid[]=
"$Header: /usr/src/local/bin/rcs/src/RCS/rcsdiff.c,v 4.6 89/05/01 15:12:27 narten Exp Locker: miyake $ Purdue CS";
#endif
/*****************************************************************************
 *                       generate difference between RCS revisions
 *****************************************************************************
 */

/* Copyright (C) 1982, 1988, 1989 Walter Tichy
   Distributed under license by the Free Software Foundation, Inc.

This file is part of RCS.

RCS is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 1, or (at your option)
any later version.

RCS is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with RCS; see the file COPYING.  If not, write to
the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.

Report problems and direct all questions to:

    rcs-bugs@cs.purdue.edu

*/




/* $Log:	rcsdiff.c,v $
 * Revision 4.6  89/05/01  15:12:27  narten
 * changed copyright header to reflect current distribution rules
 * 
 * Revision 4.5  88/11/08  12:01:51  narten
 * changes from  eggert@sm.unisys.com (Paul Eggert)
 * 
 * Revision 4.5  88/08/09  19:12:41  eggert
 * Use execv(), not system(); yield exit status like diff(1)s; allow cc -R.
 * 
 * Revision 4.4  87/12/18  11:37:46  narten
 * changes Jay Lepreau made in the 4.3 BSD version, to add support for
 * "-i", "-w", and "-t" flags and to permit flags to be bundled together, 
 * merged in.
 * 
 * Revision 4.3  87/10/18  10:31:42  narten
 * Updating version numbers. Changes relative to 1.1 actually
 * relative to 4.1
 * 
 * Revision 1.3  87/09/24  13:59:21  narten
 * Sources now pass through lint (if you ignore printf/sprintf/fprintf 
 * warnings)
 * 
 * Revision 1.2  87/03/27  14:22:15  jenkins
 * Port to suns
 * 
 * Revision 1.1  84/01/23  14:50:18  kcs
 * Initial revision
 * 
 * Revision 4.1  83/05/03  22:13:19  wft
 * Added default branch, option -q, exit status like diff.
 * Added fterror() to replace faterror().
 * 
 * Revision 3.6  83/01/15  17:52:40  wft
 * Expanded mainprogram to handle multiple RCS files.
 *
 * Revision 3.5  83/01/06  09:33:45  wft
 * Fixed passing of -c (context) option to diff.
 *
 * Revision 3.4  82/12/24  15:28:38  wft
 * Added call to catchsig().
 *
 * Revision 3.3  82/12/10  16:08:17  wft
 * Corrected checking of return code from diff; improved error msgs.
 *
 * Revision 3.2  82/12/04  13:20:09  wft
 * replaced getdelta() with gettree(). Changed diagnostics.
 *
 * Revision 3.1  82/11/28  19:25:04  wft
 * Initial revision.
 *
 */
#include <ctype.h>
#include "rcsbase.h"
#define ERRCODE 2                   /*error code for exit status            */
extern char *rindex();
#ifndef lint
static char rcsbaseid[] = RCSBASE;
#endif
static char co[] = CO;

extern int    cleanup();            /* cleanup after signals                */
extern char * mktempfile();         /*temporary file name generator         */
extern int    fterror();            /*forward for special fatal error func. */
extern struct hshentry * genrevs(); /*generate delta numbers                */
extern int    nerror;               /*counter for errors                    */
extern int    quietflag;            /*suppresses diagnostics                */
extern FILE * finptr;               /* RCS input file                       */

char *RCSfilename;
char *workfilename;
char * temp1file, * temp2file;

char bops[10];
char otherops[10];

main (argc, argv)
int argc; char **argv;
{
        char * cmdusage;
	char commarg[revlength+3];
        int  revnums;                 /* counter for revision numbers given */
        char * rev1, * rev2;          /* revision numbers from command line */
        char numericrev[revlength];   /* holds expanded revision number     */
        char * xrev1, * xrev2;        /* expanded revision numbers          */
        struct hshentry * gendeltas[hshsize];/*stores deltas to be generated*/
        struct hshentry * target;
	char * boption, * otheroption;
        int  exit_stats;
	int  diffs_found;
	char *argp;
	register c;

        catchints();
        otheroption = otherops + 2;
	boption = bops + 2;
        cmdid = "rcsdiff";
	cmdusage = "command format:\n    rcsdiff [-biwt] [-q] [-cefhn] [-rrev1] [-rrev2] file";
	diffs_found=revnums=0;
        while (--argc,++argv, argc>=1 && ((*argv)[0] == '-')) {
	    argp = &((*argv)[1]);
	    while (c = *argp++) switch (c) {
                case 'r':
		        if (*argp!='\0') {
                            if (revnums==0) {
                                    rev1= argp; revnums=1;
                            } elif (revnums==1) {
                                    rev2= argp; revnums=2;
                            } else {
				    fterror("too many revision numbers");
                            }
                        } /* do nothing for empty -r */
			argp += strlen(argp);
                        break;
                case 'b':
                case 'i':
                case 'w':
                case 't':
			if (!rindex(bops + 2, c))
				*boption++ = c;
			break;
		case 'q':
			quietflag=true;
			break;
                case 'c':
                case 'e':
                case 'f':
                case 'h':
                case 'n':
                        if (otheroption == otherops + 2) {
				*otheroption++ = c;
				if (c == 'c' && isdigit(*argp)) {
					while (isdigit(*argp) && otheroption < otherops+sizeof(otherops)-1)
						*otheroption++ = *argp++;
					if (*argp)
						faterror("-c: bad count");
					argp = "";
				}
                        } else {
				fterror("Options c,e,f,h,n are mutually exclusive");
                        }
			break;
                default:
			fterror("unknown option: %s\n%s", *argv,cmdusage);
                };
        } /* end of option processing */

	if (boption != bops + 2) {
	    bops[0] = ' ';
	    bops[1] = '-';
	    boption = bops;
	}
	if (otheroption != otherops + 2) {
 	    otherops[0] = ' ';
	    otherops[1] = '-';
	    otheroption = otherops;
	}
	if (argc<1) fterror("No input file\n%s",cmdusage);

        /* now handle all filenames */
        do {
                finptr=NULL;

                if (pairfilenames(argc,argv,true,false)!=1) continue;
		diagnose("===================================================================");
                diagnose("RCS file: %s",RCSfilename);
                if (revnums<2 && !(access(workfilename,4)==0)) {
                        error("Can't open %s",workfilename);
                        continue;
                }
                if (!trysema(RCSfilename,false)) continue; /* give up */


                gettree(); /* reads in the delta tree */

                if (Head==nil) {
                        error("no revisions present");
                        continue;
                }
                if (revnums==0)
                        rev1=Dbranch!=nil?Dbranch->num:Head->num; /* default rev1 */

                if (!expandsym(rev1,numericrev)) continue;
                if (!(target=genrevs(numericrev,(char *)nil,(char *)nil,(char *)nil,gendeltas))) continue;
                xrev1=target->num;

                if (revnums==2) {
                        if (!expandsym(rev2,numericrev)) continue;
                        if (!(target=genrevs(numericrev,(char *)nil,(char *)nil,(char *)nil,gendeltas))) continue;
                        xrev2=target->num;
                }


                temp1file=mktempfile("/tmp/",TMPFILE1);
                diagnose("retrieving revision %s",xrev1);
                VOID sprintf(commarg,"-p%s",xrev1);
                if (run((char*)nil,temp1file, co,"-q",commarg,RCSfilename,(char*)nil)){
                        error("co failed");
                        continue;
                }
                if (revnums<=1) {
                        temp2file=workfilename;
                        diagnose("diff%s%s -r%s %s",boption,otheroption,xrev1,workfilename);
                } else {
                        temp2file=mktempfile("/tmp/",TMPFILE2);
                        diagnose("retrieving revision %s",xrev2);
			VOID sprintf(commarg,"-p%s",xrev2);
                        if (run((char*)nil,temp2file, co,"-q",commarg,RCSfilename,(char *)nil)){
                                error("co failed");
                                continue;
                        }
                        diagnose("diff%s%s -r%s -r%s",boption,otheroption,xrev1,xrev2);
                }

                exit_stats =
			*boption
			? *otheroption
			  ? run((char*)nil,(char*)nil, DIFF, boption+1,	otheroption+1,	temp1file,temp2file,(char*)nil)
			  : run((char*)nil,(char*)nil, DIFF, boption+1,			temp1file,temp2file,(char*)nil)
			: *otheroption
			  ? run((char*)nil,(char*)nil, DIFF, 		otheroption+1,	temp1file,temp2file,(char*)nil)
			  : run((char*)nil,(char*)nil, DIFF, 				temp1file,temp2file,(char*)nil);

                if (exit_stats == (1 << BYTESIZ))
			diffs_found = 1;
                else if (exit_stats != 0) {
                        error ("diff failed");
                        continue;
                }
        } while (cleanup(),
                 ++argv, --argc >=1);


        exit(nerror ? ERRCODE : diffs_found);
}


/*VARARGS3*/
fterror(e, e1, e2)
char * e, * e1, * e2;
/* prints error message and terminates program with ERRCODE */
{       nerror++;
        VOID fprintf(stderr,"%s error: ",cmdid);
	VOID fprintf(stderr,e, e1, e2);
        VOID fprintf(stderr,"\n%s aborted\n",cmdid);
        VOID cleanup();
	exit(ERRCODE);
}

